src/Compilers/Core/Portable/Operations/Operation.cs

```diff
@@ -46,41 +76,203 @@ public Operation(OperationKind kind, SyntaxNode syntax, ITypeSymbol type, Option
+            (operation as Operation).SetParentOperation(parent);
+            return operation;
+        }
+
+        public static ImmutableArray<T> SetParentOperation<T>(ImmutableArray<T> operations, IOperation parent) where T : IOperation
+        {
+            // check quick bail out case first
+            if (operations.Length == 0)
+            {
+                // no element
+                return operations;
+            }
+
+            // race is okay. penalty is going through a loop one more time
+            // explicit cast is not allowed, so using "as" instead
+            // invalid expression can have null element in the array
+            if ((operations[0] as Operation)?._parentDoNotAccessDirectly != null)
+            {
+                // already initialized
+                return operations;
+            }
+
+            foreach (var operation in operations)
+            {
+                // go through slowest path
+                SetParentOperation(operation, parent);
             }
 
-            public OperationKind Kind => OperationKind.None;
+            return operations;
+        }
 
-            public SyntaxNode Syntax { get; }
+        public static T ResetParentOperation<T>(T operation) where T : IOperation
+        {
+            if (operation == null)
+            {
+                return operation;
+            }
+
+            Interlocked.Exchange(ref (operation as Operation)._parentDoNotAccessDirectly, null);
+            return operation;
+        }
 
-            public ITypeSymbol Type => null;
+        private class NoneOperation : Operation
+        {
+            private readonly Func<ImmutableArray<IOperation>> _getChildren;
 
-            public Optional<object> ConstantValue { get; }
+            public NoneOperation(SemanticModel semanticMode, SyntaxNode node, Optional<object> constantValue, Func<ImmutableArray<IOperation>> getChildren) :
+                base(OperationKind.None, semanticMode, node, type: null, constantValue: constantValue)
+            {
+                _getChildren = getChildren;
+            }
 
-            public void Accept(OperationVisitor visitor)
+            public override void Accept(OperationVisitor visitor)
             {
                 visitor.VisitNoneOperation(this);
             }
 
-            public TResult Accept<TArgument, TResult>(OperationVisitor<TArgument, TResult> visitor, TArgument argument)
+            public override TResult Accept<TArgument, TResult>(OperationVisitor<TArgument, TResult> visitor, TArgument argument)
             {
                 return visitor.VisitNoneOperation(this, argument);
             }
 
-            public IEnumerable<IOperation> Children => _getChildren().NullToEmpty();
+            public override IEnumerable<IOperation> Children
+            {
+                get
+                {
+                    foreach (var child in _getChildren().NullToEmpty().WhereNotNull())
```

Comment:
Would it be cheaper to move the NullToEmpty and WhereNotNull checks inside the loop to avoid multiple enumerators?

---
In reply to: [130171012](https://github.com/dotnet/roslyn/pull/20964#discussion_r130171012) [](ancestors = 130171012)